[macOS] WebKit should not re-export WebCore

emw-apple opened the pull request:
#### e52304a6c26afa75abf5ae47602516556bcc4ade
<pre>
[macOS] WebKit should not re-export WebCore
<a href="https://bugs.webkit.org/show_bug.cgi?id=259352">https://bugs.webkit.org/show_bug.cgi?id=259352</a>
rdar://110785133

Reviewed by Alexey Proskuryakov.

Stop re-exporting WebCore&apos;s symbols in WebKit on macOS. This makes Mac
WebKit behave more like other Darwin-based platforms: WebKit.framework
re-exports all public symbols from WebKitLegacy, and only select symbols
from WebCore.

WebKitLegacy&apos;s export lists (Source/WebKitLegacy/mac/*.exp) contain the
WebCore symbols still visible to WebKit clients. And now that WebCore is
not implicitly visible to WebKit clients, some additional WebCore
symbols can be exported for testing.

Additionally, the migrate-header-rule re-export logic--which ensures
that all symbols corresponding to migrated headers are exported--is now
used on macOS (previously, it was on for every platform *but* macOS).
This does not actually expose new symbols to be exported, but ensures
that any new SPI implemented in WebCore will be exported as before.

* Source/WebCore/Configurations/WebCore.xcconfig:
  - Replace &quot;-umbrella WebKit&quot; with &quot;-allowable_client WebKit&quot;. The
    former implies that WebCore is re-exported from WebKit.
  - Make testing tools allowable clients on all platforms, not just
    embedded.

* Source/WebCore/PAL/pal/cocoa/DataDetectorsCoreSoftLink.mm: Export some
  DataDetectors soft-link calls used by WebCoreTestSupport.

* Source/WebKit/Configurations/WebKit.xcconfig: Remove WebCore from
  REEXPORTED_FRAMEWORK_NAMES.
* Source/WebKitLegacy/WebKitLegacy.xcodeproj/project.pbxproj: Update
  &quot;Generate Export Files&quot; script phase to use the results of
  migrate-header-rule on all platforms.
* Source/WebKitLegacy/scripts/migrate-header-rule: Generate an export
  file on all platforms.

* Tools/DumpRenderTree/mac/Configurations/DumpRenderTree.xcconfig: Link
  against WebCore.
* Tools/TestWebKitAPI/Configurations/TestWebKitAPI.xcconfig: Link
  against WebCore.
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj: Link
  against WebCore in TestWebKitAPI.wkbundle.
* Tools/WebKitTestRunner/WebKitTestRunner.xcodeproj/project.pbxproj:
  Link against WebCore.

Canonical link: <a href="https://commits.webkit.org/266838@main">https://commits.webkit.org/266838@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a0a26b288e45e930d0bcd9d30a6d399c52e07468

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14907 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15211 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15565 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16655 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/14024 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/15044 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17732 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15312 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/16659 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15087 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15569 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12662 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17388 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12848 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/13448 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/20414 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13927 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13616 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16850 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14166 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/11975 "Passed tests") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/13451 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3598 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17783 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14009 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Source/WebKitLegacy/WebKitLegacy.xcodeproj/project.pbxproj:
```diff
@@ -3184,7 +3184,7 @@
 			runOnlyForDeploymentPostprocessing = 0;
 			shellPath = /bin/sh;
-			shellScript = "set -e\n\nEXP_GENERATED=\"${SCRIPT_OUTPUT_FILE_0}\"\n\n[[ -z \"${EXP_GENERATED}\" ]] && exit 1\n\nif [[ \"${WK_PLATFORM_NAME}\" != \"macosx\" ]]; then\n    EXP_WEBKIT_PLATFORM=\"${SCRIPT_INPUT_FILE_2}\"\nelse\n    EXP_WEBKIT_PLATFORM=\"${SCRIPT_INPUT_FILE_1}\"\nfi\n\ncat \"${SCRIPT_INPUT_FILE_0}\" > \"${EXP_GENERATED}\"\ncat \"${EXP_WEBKIT_PLATFORM}\" >> \"${EXP_GENERATED}\"\n\n ...
+			shellScript = "set -e\nexec > \"${SCRIPT_OUTPUT_FILE_0}\"\n\ncat \"${SCRIPT_INPUT_FILE_0}\"\n\nif [[ \"${WK_PLATFORM_NAME}\" == \"macosx\" ]]; then\n    cat \"${SCRIPT_INPUT_FILE_1}\"\nelse\n    cat \"${SCRIPT_INPUT_FILE_2}\"\nfi\n\necho\necho \"# Reexported from migrated WebCore headers\"\ncat \"${SCRIPT_INPUT_FILE_3}\"\n\nif [[ \"${WK_BUILD_FOR_TESTING}\" == \"YES\" ]]; then\n    cat \"${SCR ...
```
Diff of the script itself:

```diff
--- A/app.kaleidoscope.kaleidoscope_dynamic_file_4BF6009F-2A93-4A82-812A-3DBB6F39D537-94887-000005E84ECEF7ED.txt
+++ B/app.kaleidoscope.kaleidoscope_dynamic_file_E99C066C-6D3B-41E8-99D3-92246F71685C-94887-000005E85C3BB80B.txt
@@ -1,24 +1,18 @@
 set -e
+exec > "${SCRIPT_OUTPUT_FILE_0}"
 
+cat "${SCRIPT_INPUT_FILE_0}"
-EXP_GENERATED="${SCRIPT_OUTPUT_FILE_0}"
 
+if [[ "${WK_PLATFORM_NAME}" == "macosx" ]]; then
+    cat "${SCRIPT_INPUT_FILE_1}"
-[[ -z "${EXP_GENERATED}" ]] && exit 1
-
-if [[ "${WK_PLATFORM_NAME}" != "macosx" ]]; then
-    EXP_WEBKIT_PLATFORM="${SCRIPT_INPUT_FILE_2}"
 else
+    cat "${SCRIPT_INPUT_FILE_2}"
-    EXP_WEBKIT_PLATFORM="${SCRIPT_INPUT_FILE_1}"
 fi
 
+echo
+echo "# Reexported from migrated WebCore headers"
+cat "${SCRIPT_INPUT_FILE_3}"
-cat "${SCRIPT_INPUT_FILE_0}" > "${EXP_GENERATED}"
-cat "${EXP_WEBKIT_PLATFORM}" >> "${EXP_GENERATED}"
-
-if [ "${WK_PLATFORM_NAME}" != "macosx" ]; then
-    echo >> "${EXP_GENERATED}"
-    echo "# Reexported from migrated WebCore headers" >> "${EXP_GENERATED}"
-    cat "${SCRIPT_INPUT_FILE_3}" >> "${EXP_GENERATED}"
-fi
 
 if [[ "${WK_BUILD_FOR_TESTING}" == "YES" ]]; then
+    cat "${SCRIPT_INPUT_FILE_4}"
-    cat "${SCRIPT_INPUT_FILE_4}" >> "${EXP_GENERATED}"
 fi
```

I refactored it a bit to avoid having to put shell redirections on every line.

aproskuryakov reviewed (approved):
It's curious that `ReexportedFromWebCore.exp` is identical between Mac and iOS.

> And now that WebCore is not implicitly visible to WebKit clients, some additional WebCore
symbols can be exported for testing.

I do not see what this refers to. Test tools link to WebCore now, what are the additional things being exported?

> Export some DataDector soft-link calls used by WebCoreTestSupport.

Aren't these symbols used by WebKit proper, too? Also, typo in `DataDetectors`.

In fact, WebCoreTestSupport may not need soft-linking these at all. The common reasons for soft-linking are fallback for when dependency is missing in Recovery (doesn't apply to test tools), and breaking build time dependency cycles (also doesn't apply to test tools, and AFAIK there are better ways to solve this case anyway).

emw-apple committed:
[macOS] WebKit should not re-export WebCore
https://bugs.webkit.org/show_bug.cgi?id=259352
rdar://110785133

Reviewed by Alexey Proskuryakov.

Stop re-exporting WebCore's symbols in WebKit on macOS. This makes Mac
WebKit behave more like other Darwin-based platforms: WebKit.framework
re-exports all public symbols from WebKitLegacy, and only select symbols
from WebCore.

WebKitLegacy's export lists (Source/WebKitLegacy/mac/*.exp) contain the
WebCore symbols still visible to WebKit clients. And now that WebCore is
not implicitly visible to WebKit clients, some additional WebCore
symbols can be exported for testing.

Additionally, the migrate-header-rule re-export logic--which ensures
that all symbols corresponding to migrated headers are exported--is now
used on macOS (previously, it was on for every platform *but* macOS).
This does not actually expose new symbols to be exported, but ensures
that any new SPI implemented in WebCore will be exported as before.

* Source/WebCore/Configurations/WebCore.xcconfig:
  - Replace "-umbrella WebKit" with "-allowable_client WebKit". The
    former implies that WebCore is re-exported from WebKit.
  - Make testing tools allowable clients on all platforms, not just
    embedded.

* Source/WebCore/PAL/pal/cocoa/DataDetectorsCoreSoftLink.mm: Export some
  DataDetectors soft-link calls used by WebCoreTestSupport.

* Source/WebKit/Configurations/WebKit.xcconfig: Remove WebCore from
  REEXPORTED_FRAMEWORK_NAMES.
* Source/WebKitLegacy/WebKitLegacy.xcodeproj/project.pbxproj: Update
  "Generate Export Files" script phase to use the results of
  migrate-header-rule on all platforms.
* Source/WebKitLegacy/scripts/migrate-header-rule: Generate an export
  file on all platforms.

* Tools/DumpRenderTree/mac/Configurations/DumpRenderTree.xcconfig: Link
  against WebCore.
* Tools/TestWebKitAPI/Configurations/TestWebKitAPI.xcconfig: Link
  against WebCore.
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj: Link
  against WebCore in TestWebKitAPI.wkbundle.
* Tools/WebKitTestRunner/WebKitTestRunner.xcodeproj/project.pbxproj:
  Link against WebCore.

Canonical link: https://commits.webkit.org/266838@main
